src/metabase/query_processor/annotate.clj

```diff
@@ -371,13 +371,17 @@
   2.  Resolves the Fields returned in the results and adds information like `:columns` and `:cols` expected by the
       frontend."
   [query {:keys [columns rows], :as results}]
-  (let [row-maps (for [row rows]
-                   (zipmap columns row))
-        cols    (resolve-sort-and-format-columns (:query query) (distinct columns) (take 10 row-maps))
-        columns (mapv :name cols)]
+  (let [initial-rows      (for [row (take 10 rows)]
+                            (zipmap columns row))
+        cols              (resolve-sort-and-format-columns (:query query) (distinct columns) initial-rows)
+        original-ordering (zipmap columns (range))
+        columns           (mapv :name cols)
+        reordering        (map original-ordering columns)]
```

Comment:
Although I knew what this code was doing by reading the description in the PR, I had to reread this code several times before I understood it. I where I was struggling was with the names in the `let` I think. Improving those might help. Might make it easier to understand if that was refactored to a function also. One thing that came to mind was what if there was a function `pre-sort-index->post-sort-index` (or similar), that returned an int->int map or something similar.